File: Algorithm/Python/173. Binary Search Tree Iterator.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-07-16 09:02:25
# @Last modified by:   WuLC
# @Last Modified time: 2016-07-16 09:34:54
# @Email: liangchaowu5@gmail.com

# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# method 1, O(n) space
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.curr_node = root
        self.stack = []
        self.values = []
        while len(self.stack)>0 or self.curr_node:
            if self.curr_node:
                self.stack.append(self.curr_node)
                self.curr_node = self.curr_node.left
            else:
                self.curr_node = self.stack.pop()
                self.values.append(self.curr_node.val)
                self.curr_node = self.curr_node.right
        self.values.reverse()
            
        
    def hasNext(self):
        """
        :rtype: bool
        """
        return len(self.values)>0 

    def next(self):
        """
        :rtype: int
        """
        return self.values.pop()
        

# method 2, O(h) space
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = []
        self.curr = root
        
    def hasNext(self):
        """
        :rtype: bool
        """
        return len(self.stack)>0 or self.curr

    def next(self):
        """
        :rtype: int
        """
        while self.curr:
            self.stack.append(self.curr)
            self.curr = self.curr.left
        tmp = self.stack.pop()
        self.curr = tmp.right
        return tmp.val

File: Algorithm/Java/173. Binary Search Tree Iterator.java
/*
* @Author: WuLC
* @Date:   2017-03-01 09:25:34
* @Last Modified by:   WuLC
* @Last Modified time: 2017-03-01 09:26:28
* @Email: liangchaowu5@gmail.com
*/

/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */



// traverse the tree in the order right-mid-left and store them in a tree 
public class BSTIterator 
{
    private Stack<Integer> values;
    public BSTIterator(TreeNode root) 
    {
        values = new Stack<Integer>();
        Stack<TreeNode> nodes = new Stack<TreeNode>();
        TreeNode curr = root;
        while(nodes.size()>0 || curr!=null)
        {
            if (curr != null) 
            {
                nodes.push(curr);
                curr = curr.right;
            }
            else
            {
                curr = nodes.pop();
                values.push(curr.val);
                curr = curr.left;
            }
        }
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() 
    {
        return values.size() != 0;
    }

    /** @return the next smallest number */
    public int next() 
    {
        return values.pop();
    }
}